tpm2/seal.go

```diff
@@ -0,0 +1,445 @@
+	// random 32-byte number.
+	// If not set, one is generated automatically.
+	AuthKey secboot.AuxiliaryKey
+
+	// AuthorizedSnapModels is a list of models initially authorized to access
+	// the data protected by the newly created key. These can be updated later
+	// on by calling secboot.KeyData.SetAuthorizedSnapModels.
+	AuthorizedSnapModels []secboot.SnapModel
+}
+
+type keyDataPolicyParams struct {
+	Alg    tpm2.HashAlgorithmId
+	Data   keyDataPolicy
+	Digest tpm2.Digest
+}
+
+// makeKeyDataWithPolicy protects the supplied keys using the supplied keySealer and
+// policy data. This can be called multiple times to protect an arbitary number of
+// keys with an identical policy.
+func makeKeyDataWithPolicy(key secboot.DiskUnlockKey, authKey secboot.AuxiliaryKey, policy *keyDataPolicyParams, sealer keySealer) (*secboot.KeyData, error) {
+	var symKey [32 + aes.BlockSize]byte
+	if _, err := rand.Read(symKey[:]); err != nil {
+		return nil, xerrors.Errorf("cannot create symmetric key: %w", err)
+	}
+
+	priv, pub, importSymSeed, err := sealer.CreateSealedObject(symKey[:], policy.Alg, policy.Digest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new SealedKeyObject.
+	data, err := newKeyData(priv, pub, importSymSeed, policy.Data)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create key data: %w", err)
+	}
+	skd := &SealedKeyData{sealedKeyDataBase: sealedKeyDataBase{data: data}}
+
+	// Create encrypted payload
+	payload := secboot.MarshalKeys(key, authKey)
+
+	b, err := aes.NewCipher(symKey[:32])
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create new cipher: %w", err)
+	}
+	stream := cipher.NewCFBEncrypter(b, symKey[32:])
+	stream.XORKeyStream(payload, payload)
+
+	kd, err := secbootNewKeyData(&secboot.KeyParams{
+		Handle:           skd,
+		EncryptedPayload: payload,
+		PlatformName:     platformName,
+		AuxiliaryKey:     authKey,
+		// Hardcode SHA-256 here. We already hardcode this as the name algorithm
+		// for the sealed object and elliptic key.
+		SnapModelAuthHash: crypto.SHA256})
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create key data object: %w", err)
+	}
+
+	return kd, nil
+}
+
+type createdPcrPolicyCounter struct {
+	tpm     *tpm2.TPMContext
+	session tpm2.SessionContext
+	pub     *tpm2.NVPublic
+}
+
+func (c *createdPcrPolicyCounter) Pub() *tpm2.NVPublic {
+	if c == nil {
+		return nil
+	}
+	return c.pub
+}
+
+func (c *createdPcrPolicyCounter) undefineOnError(err error) {
+	if c == nil {
+		return
+	}
+
+	if err == nil {
+		return
+	}
+
+	index, err := tpm2.CreateNVIndexResourceContextFromPublic(c.pub)
+	if err != nil {
+		return
+	}
+	c.tpm.NVUndefineSpace(c.tpm.OwnerHandleContext(), index, c.session)
+}
+
+// makeKeyDataPolicy creates the policy data required to seal a key with makeKeyDataWithPolicy
+// and creates a PCR policy counter if required.
+func makeKeyDataPolicy(tpm *tpm2.TPMContext, pcrPolicyCounterHandle tpm2.Handle, authKey secboot.AuxiliaryKey,
```

Comment:
this should probably be closer to keyDataPolicyParams definition, the latter name maybe needs tweaks